util/bloom/xor.go

```diff
@@ -0,0 +1,259 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package bloom
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+
+	"github.com/algorand/xorfilter"
+)
+
+// XorBuilder is a local alias for xorfilter.Builder
+type XorBuilder = xorfilter.Builder
+
+// XorFilter is a faster more efficient alternative to a Bloom filter
+// An XorFilter object can be used as is or with optional adittional setup.
+type XorFilter struct {
+	xor     *xorfilter.Xor32
+	holding []uint64
+
+	b *XorBuilder
+}
+
+// NewXor returns an XorFilter with an internal map created with a size hint and an optional *XorBuilder (may be nil)
+// The Builder is not thread safe and should only be used by one thread at a time.
+func NewXor(hint int, builder *XorBuilder) *XorFilter {
```

Comment:
Hm, It used to be. This could be cleaned up. TODO

Reply:
I think it should be used to pre-allocate `holding`. Otherwise, we'll have O(n^2) in
```golang
xf.holding = append(xf.holding, k)
}
```

Reply:
`append()` will allocate a new array that is a constant times bigger than previous, so it's still O(n), right? That's what `std::vector` in C++ does.

Reply:
That's not exactly the way append work, but even if so, then it would allocate at:
1+2+4+8+...n. and on each allocation, it would copy the entire backing arry into the new allocation. So, the last append would copy n/2 entries, the one before would do n/4, and so forth. That makes it into n×logn, assuming the constant you mentioned above. And surprise! Append doesnt use a constant.. which makes it much worse.

Reply:
1+2+4+8+...+2^k = 2^(k+1) - 1, so O(n). Where did you see information that append doesn't use a constant. This article says it does. https://medium.com/vendasta/golang-the-time-complexity-of-append-2177dcfb6bad

Reply:
that's not true : https://play.golang.org/p/LTn9FzxAzt8

Reply:
I ran the example. Like the article says, below size 1024 the constant factor is 2, after - 1.25.

Reply:
I updated the code to use hint as a capacity for allocation of holding.